firedrake/solving.py

```diff
@@ -215,6 +218,19 @@ def _la_solve(A, x, b, **kwargs):
 
     bcs, solver_parameters, nullspace, nullspace_T, near_nullspace, \
         options_prefix = _extract_linear_solver_args(A, x, b, **kwargs)
+
+    from firedrake.bcs import EquationBC
+
+    def rassert_linear(bcs):
+        for bc in bcs:
+            if isinstance(bc, EquationBC):
+                if bc.is_linear:
+                    rassert_linear(bc.bcs)
+                else:
+                    raise RuntimeError("EquationBCs must also be linear when solving linear system.")
```

Comment:
Again, this looks like if you could iterate over the bcs of an `EquationBC` sanely, then you wouldn't have to hand-craft this recursion. Perhaps it it implemented:

```
class DirichletBC(...):
    is_linear = True
    def __iter__(self):
        yield self

class EquationBC(...):
    def __iter__(self):
        yield self
        yield from itertools.chain(*self.bcs)
```

Then this code would be something like:

```
bcs = _extract_bcs(bcs)
assert all(bc.is_linear for bc in bcs), "..."
```

Reply:
This is cool.  Added __iter__ in `BCBase` and in `NonlinearVariationalProblem` and removed recursions wherever we can "flatten" bcs.